RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,358 @@
+2. Preserve `torch.Tensor` subclasses when calling `torch` functions on them
+3. Use the PyTorch API with `torch.Tensor`-like objects that are _not_ `torch.Tensor`
+   subclasses
+4. Preserve `torch.Tensor` subclasses when calling `torch.Tensor` methods.
+5. Propagating subclass instances correctly also with operators, using
+   views/slices/indexing/etc.
+6. Preserve subclass attributes when using methods or views/slices/indexing.
+7. A way to insert code that operates on both functions and methods uniformly
+   (so we can write a single function that overrides all operators).
+8. The ability to give external libraries a way to also define
+   functions/methods that follow the `__torch_function__` protocol.
+
+Goals 1‒6 are explicitly about subclassing, goal 7 is already partially achieved via the `__torch_function__` protocol (which we're proposing to extend to methods), and goal 8 is a by-product required to make overridden `torch.Tensor` subclass methods behave similar to `torch.Tensor` methods.
+
+Achieving interoperability with NumPy and adopting its array protocols is out
+of scope for this proposal and we propose to defer it to a later proposal.
+
+We propose to solve this problem with the following changes to PyTorch:
+
+1. Make methods and operators of `torch.Tensor` go through the
+   `__torch_function__` machinery.
+2. Add a `types` argument to `__torch_function__`, to make it match NumPy's
+   `__array_function__`.
+3. Make `torch.Tensor._make_subclass` public API by renaming it to `torch.Tensor.make_subclass`.
+4. Make `torch.Tensor` gain a generic implementation of `__torch_function__`.
+
+## Usage and Impact
+Once this proposal is merged, users of subclasses of `torch.Tensor` will have
+a much more streamlined experience. Namely, the following code example will
+work as-is, without the need for any further modification:
+
+```python
+class SubTensor(torch.Tensor):
+    a = 1
+
+t = SubTensor([1])
+s = t.sum()
+isinstance(s, SubTensor)  # True
+s.a  # 1
+i = t[0]
+isinstance(i, SubTensor)  # True
+i.a  # 1
+
+s2 = t + torch.Tensor(1)
+isinstance(s2, SubTensor)  # True
+s2.a  # 1
+
+s3 = torch.Tensor(1) + t
+isinstance(s3, SubTensor)  # True
+s3.a  # 1
+```
+
+Additionally, it will provide subclass authors the ability to also modify the
+results of methods and operators in `__torch_function__`, along with regular
+function calls, and to modify the result to their specific use-case, perform
+logging, or otherwise change the result or the action of the method. For
+example:
+
+```python
+import logging
+
+class LoggingTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        logging.info(f"func: {func.__name__}, args: {args!r}, kwargs: {kwargs!r}")
+        return super().__torch_function__(
+            func,
+            tuple(
+                t if not issubclass(t, LoggingTensor) else torch.Tensor for t in types
+            ),
+            args,
+            kwargs
```

Comment:
I think there is a teaching moment here. Why should we call `super().__torch_function__` rather than directly calling `func` which is the backend implementation? Answer: if we called func directly, we wouldn't wrap the resulting tensor in a LoggingTensor (thus losing the Logging-ness for downstream calls). This could be a valid choice for some kinds of tensors! But not for this one, probably.

Reply:
Actually, it seems like calling `func` would not work at all, if we follow the calling convention from existing `__torch_function__`. The existing calling convention is that `func` is exactly something like `torch.Tensor.__add__`, which is a *dispatching* function; if you called `func` directly you would infinite loop, if `func` is a function that goes to the dispatcher, which would in turn call `__torch_function__` again (infinite loop). This kind of suggests we need either (1) a way to name functions that bypasses dispatch, or (2) a way to invoke a function without triggering dispatch. I'm guessing this is what the `_implementation` field on these functions does; worth calling this out explicitly in the RFC!

Reply:
@hameerabbasi capturing the above comments about "teaching moment" in a code comment would be useful.

> This kind of suggests we need either (1) a way to name functions that bypasses dispatch, or (2) a way to invoke a function without triggering dispatch. I'm guessing this is what the `_implementation` field on these functions does; worth calling this out explicitly in the RFC!

So you did this by providing a (non-public?) context manager, right @hameerabbasi?

Reply:
That's correct, that's the conclusion we came to in a call with @ezyang 

The upshot of it was that we're not overly concerned about thread forks, the state of `__torch_function__` can be captured within a thread-local variable, and there aren't any forks between the moment the function is called and dispatching, where `__torch_function__` is checked.

This is also the reason it's private -- the context doesn't survive when new threads are spawned. pytorch/pytorch#28520 may resolve this.